packages/react-utilities/src/compose/types.ts

```diff
@@ -1,45 +1,95 @@
 import * as React from 'react';
 
-/**
- * Generic name to any dictionary.
- */
+// TODO: delete GenericDictionary, this is harmful for overall typings of the lib
 // eslint-disable-next-line @typescript-eslint/no-explicit-any
 export type GenericDictionary = Record<string, any>;
 
 /**
- * Class dictionary.
+ * Generic record of possible ShorhandProps
  */
-export type ClassDictionary = Record<string, string>;
+// eslint-disable-next-line @typescript-eslint/no-explicit-any
+export type SlotPropsRecord = Record<string, Record<string, any> | undefined>;
+
+export type ShorthandPropsRecord<Record extends SlotPropsRecord = {}> = {
+  [K in keyof Record]: ShorthandProps<NonNullable<Record[K]>>;
+};
+
+export type ObjectShorthandPropsRecord<Record extends SlotPropsRecord = {}> = {
+  [K in keyof Record]: ObjectShorthandProps<NonNullable<Record[K]>>;
+};
+
+export type ShorthandRenderFunction<Props> = (Component: React.ElementType<Props>, props: Props) => React.ReactNode;
+
+export type ShorthandProps<Props = {}> =
+  | React.ReactChild
+  | React.ReactNodeArray
+  | React.ReactPortal
+  | number
+  | null
+  | undefined
+  | ObjectShorthandProps<Props>;
+
+export type ObjectShorthandProps<Props extends { children?: React.ReactNode } = {}> = Props &
+  Pick<ComponentProps, 'as'> & {
+    children?: Props['children'] | ShorthandRenderFunction<Props>;
+  };
+
+export type ComponentProps<Record extends SlotPropsRecord = {}> = {
+  as?: keyof JSX.IntrinsicElements;
+  className?: string;
```

Comment:
do we need `className` here ? it generally there by extending `React.HTMLAttributes`

Reply:
This is inhereted by the previous implementation, didn't want to create too possible breaking changes.

Reply:
I don't see why that would matter, especially if there's already a `compat` version of these typings